dlt/common/runtime/telemetry.py

```diff
@@ -112,6 +114,65 @@ def _track(success: bool) -> None:
                 _track(False)
                 raise
 
-        return _wrap  # type: ignore
+        return cast(TFun, _wrap)
+
+    return decorator
+
+
+def with_dataset_access_telemetry() -> Callable[[TFun], TFun]:
+    """Track dataset access but only once per pipeline instance"""
+
+    def decorator(f: TFun) -> TFun:
+        def _wrap(*f_args: Any, **f_kwargs: Any) -> Any:
+            pipeline = f_args[0] if f_args else None
+            schema: Union[Schema, str, None] = f_kwargs.get("schema") or (
+                f_args[1] if len(f_args) > 1 else None
+            )
+
+            def _track_dataset_access_once(success: bool) -> None:
+                if pipeline and not getattr(pipeline, "_dataset_access_tracked", False):
+                    with contextlib.suppress(Exception):
```

Comment:
OK I think we do not really need decorator at all. this is reproducing too much of logic that is in `def dataset(self, schema: Union[Schema, str, None] = None)` method.

my take: just send event after:
```py
dlt.dataset(
            self._destination,
            self.dataset_name,
            schema=schema,
        )
```
decorator is good if we plan to decorate many functions with it. right now it is just one function and it won't generalize to anything else.

Reply:
also: extract tracking itself to `track.py` (creating and sending props). look at `on_end_trace_step`: we have code duplication